### DISCUSSION TITLE
Types: String result scrambled | exception with uint64 | question to bit and byte

### DISCUSSION THREAD

**Discussion by gutschein** [aldas/modbus-tcp-client#84] [Q&A] [answered]
Now I'm working on different types. int/uint16/32 are working fine. But I've some questions:

- With type=string I get "scrambled" results. Is there something to do or a parameter to get the expected result?

Example: expected MAC: 
`00:40:AD:91:D2:8B`

via modbus using "type":"string","length":"32" (same result with "48" or "64")
`004::0DA9::12D8:`

1 byte (here: "B" from end) is missing and wrong order. Same result with BIG_ENDIAN and BIG_ENDIAN_LOW_WORD_FIRST. 

- With uint64 and int64 I get an exception error:
```
.../modbus-tcp-client/vendor/aldas/modbus-tcp-client/src/Utils/Types.php | Line: 335
"64-bit PHP supports only up to 63-bit signed integers. Current input has 64th bit set and overflows. Hex: ffffffffffffffff"
```

- I'm not sure that I understand type = bit: Do I get the information 0/1 of a dedicated 1 Bit at the address byte? What is the value range for the attribute "bit" for type=bit? In your example you used "15" - but has a Byte not just 8 bits? Probably I missunderstood your example "->bit(256, 15, 'pump2_feedbackalarm_do')" and my devices do not have such types, but I like to understand...

- What are the 2 allowed boolean values for attribute "firstByte" for type=byte? Just to be sure

Thank you!

**Comment by aldas** [comment]
-----------------

> With type=string I get "scrambled" results

For string default (global) endian is used to extract value. This meant for your case that BIG_ENDIAN_LOW_WORD_FIRST was used. Characters are in weird order (reverse pairs) because of  big-endian and little endian are bytes in reverse order when compared and  `low word first` means that byte range you are asking consist of register (2 byte pairs) and those registers are reverse order - low word/register is first.

You might want to read this https://www.simplymodbus.ca/FAQ.htm#Order

I have added way to set endian when extracting strings with v2.3.0 See https://github.com/aldas/modbus-tcp-client/blob/1fb85de630c45154c015f779c0785987d54c70e1/src/Composer/Read/ReadRegistersBuilder.php#L151 

You can try LITTLE_ENDIAN to read that string - probably it will give better results.

-----------------

> With uint64 and int64 I get an exception error:

As it says. PHP can not handle all uint64 values. Change your data type to int64 and if your values are not extremely large you are fine. NB: `ffffffffffffffff` is max value for (u)int64 - are your device using that value to indicate something?

NB: if you really need to use uint64 and that value is contant problem you could set `errorCallback` https://github.com/aldas/modbus-tcp-client/blob/1fb85de630c45154c015f779c0785987d54c70e1/src/Composer/Read/ReadRegistersBuilder.php#L103  for that address and catch that extraction error and return something else for that address. 

This test shows usage:
https://github.com/aldas/modbus-tcp-client/blob/1fb85de630c45154c015f779c0785987d54c70e1/tests/unit/Composer/Read/Register/ReadRegisterAddressTest.php#L104

-----------------

> I'm not sure that I understand type = bit: 

Bit returns N-th bit of returned register. Each register is 2 bytes (16 bits).  This is handy when your device memory for `coils` and `register` are overlapping. Then you could use FC3 etc calls to read coils (bits) values along with register values and save resources by not sending extra FC1/FC2 requests.

Byte allows you to return first or second byte of a register (2 byte value).. for example if you store uint8 value in register.

**Reply by gutschein** [reply]
**STRING**
Library: It was unexpected for me, that the endian of strings are not using the global variable for setting the endian.
Device: If my device uses another endian as for other values it would be unexpected, too.

> I have added way to set endian when extracting strings with v2.3.0 See

Sounds good that you offer now an endian per string register. But should it not use the endian of the default variable at first? Maybe there is no need to set a specific endian for a string in the most cases...

Anyway...I tried it with using 'endian" in the array (like "length"), but it doesn't work. 
```
        (
            [address] => 40497
            [name] => MAC
            [type] => string
            [length] => 32
            [endian] => BIG_ENDIAN
            [uri] => 192.168.xxx.yyy:502
            [unitId] => 3
        )
```
Probably I missunderstood how to use it properly.

**UINT64**
check. indeed...int64 should be enough for the most usecases... 🤣  my device has just 3 of this registers and they are all -1. That's fine for me. Thank you

**BIT**
check. with 2 bytes "15" makes sense. Even if my device doesn't uses bit I understand the usecase. Thank you.

**BYTE**
I've still my question... 😄 
What are the 2 allowed boolean values for attribute "firstByte" for type=byte? Just to be sure how to use...

**Comment by aldas** [comment]
> STRING 

Are you passing endian value as string or int? It should be int. 

> BYTE

true/false, and if you send some string usual string to int PHP conversion rules apply.


p.s. 

You really should use debugger to see what code does (step through each executed line of code by line by line). 

If you are using VS Code then see

* https://code.visualstudio.com/docs/editor/debugging
* https://code.visualstudio.com/docs/languages/php#_debugging
* https://www.cloudways.com/blog/php-debug/
* https://youtu.be/LNIvugvmCyQ

**Comment by gutschein** [comment, marked as answer]
I wondered two times..
* I used string as I used the it for defining the global variable. I didn't expect to use here something else... but with INT it works 😄 
* The docs of my device tells me "BIG_ENDIAN" at all - but with 2 (= LITTLE_ENDIAN) the strings work... 😄 

Results contains now...
`["MAC"] => 00:40:AD:91:D2:8B`

Thank you, Martti! Perfect now! 😄 

If somebody else comes to this pitfall, too... here are the values:
1 = BIG_ENDIAN
5 = BIG_ENDIAN_LOW_WORD_FIRST
2 = LITTLE_ENDIAN
6 = LITTLE_ENDIAN_LOW_WORD_FIRST

**Comment by gutschein** [comment]
...and thank you for your coding advices... by now just nano and `echo/print_r`-try&error. I should do better...
